### ISSUE TITLE
Reboot held up by Crowdsec refusing to stop

### ISSUE THREAD

**Issue by roens** [opnsense/plugins#4262]
**Important notices**
Before you add a new report, we ask you kindly to acknowledge the following:

- [x] I have read the contributing guide lines at https://github.com/opnsense/plugins/blob/master/CONTRIBUTING.md
- [x] I have searched the existing issues, open and closed, and I'm convinced that mine is new.
- [x] The title contains the plugin to which this issue belongs

**Describe the bug**
I've lost track of which release this started with (for me). I regret not bringing this up when it began. But the trouble is that rebooting OPNsense is held up by Crowdsec "refusing" to quit. [Others](https://forum.opnsense.org/index.php?topic=41667.0) have [faced this](https://forum.opnsense.org/index.php?topic=43067.0) this [as well](https://forum.opnsense.org/index.php?topic=34355.0).

Specifically (and for me, most commonly), this occurs for updates requiring reboot. The update log reported to the webUI ends with something like:

```
[39/48] Upgrading crowdsec from 1.6.2_4 to 1.6.3_1...
[39/48] Extracting crowdsec-1.6.3_1: .......... done
crowdsec is running as pid 84792.
Stopping crowdsec.
Waiting for PIDS: 84792
```

It will hang there until I ssh in, and kill that PID, resulting in it proceeding with the updates and automated reboot.

**To Reproduce**
1. Install an update which requires reboot OR choose: Power -> Reboot -> Yes
2. Expect it to reboot: find it is not
3. Kill the Crowdsec processes:
   - `pkill -9 -f 'daemon: crowdsec'`
   - `pkill -9 -f 'crowdsec -c'` (this PID may also need to be whacked, but not always)
5. Not that the reboot proceeds (if held up for an update, the update log in webUI will resume progressing)
   - … though sometimes it still won't and additionally requires a `reboot` from that same ssh session
   - OTOH, if a `reboot` is attempted prior to killing Crowdsec, that will fail for waiting for Crowdsec to quit nicely, requiring step 3 above

**Expected behavior**
That a reboot triggered will have no problem killing/quitting all running processes, and finally resulting in rebooting the system.

**Screenshots**
n/a

**Relevant log files**
(none that I could find, aside from update log in webUI pasted above)

**Additional context**
Once this behavior started, it has been consistently like this, every time.

**Environment**

OPNsense 24.7.4_1-amd64
FreeBSD 14.1-RELEASE-p4
OpenSSL 3.0.15

System: Supermicro SYS-E302-9D
CPU: Intel(R) Xeon(R) D-2123IT CPU @ 2.20GHz
Network: Intel I350-AM4 & Intel X557

**Comment by Starkstromkonsument** [comment]
Thanks for creating this issue. Same Problem here. It also happens if you try to stop or restart Crowdsec via the WebUI.

IMHO there are two Problems:

1. The Crowdsec stop routine does not work properly
2. OPNsense does not kill hanging processes during shutdown / reboot after a reasonable timeout --> should this be a separate issue at opnsense/core?

**Comment by fichtner** [comment]
I don’t agree with no 2. this is a sporadic issue that doesn’t have operational impact overall (as in breaks running operation of the firewall).


Cheers,
Franco

**Comment by davesnthere** [comment]
Having the same issue here too. 
After the hotfix update (24.7.5_3) OPNsense will not reboot. 
I tried stopping from the GUI but after a couple min, no change.
Tried disabling the first 3 option in the CrowdSec settings then attempt to stop service, no luck.
Just discovered, for what ever reason, I can not ssh into my system.
Will have to wait till I get home and on the local console to try pkill.

**Comment by kumy** [comment]
Facing the same issue. Anything you wanna test?

**Comment by fichtner** [comment]
Crowdsec made a barely visible effort here https://www.reddit.com/r/opnsense/comments/1fpt7xn/comment/lp3lixx/ -- but it would be nice to have someone on GitHub like in the old days ;)

**Comment by RudiKlein** [comment]
@fichtner I have been coping with this sporadic issue every upgrade. Today again.
Maybe lots of people are having this issue but use the workaround without reporting it.
It's not a big issue, just very annoying.

bye,

Rudi

**Comment by fichtner** [comment]
I agree and hope crowdsec fill fix the issue soon in their code.

**Comment by Unspec7** [comment]
I too am facing this issue.

**Comment by nhatlinh1982** [comment]
I am too facing this issue. Not sure how long it has been having since i do not reboot my OPNsense too often

**Comment by daygle** [comment]
Same issue here too. In fact I have been experiencing for some time :(

**Comment by mmetc** [comment]
Hi,

could you test this

```
# fetch -o /usr/local/etc/rc.d/crowdsec https://github.com/crowdsecurity/plugins/releases/download/crowdsec-1.6.3-2-hotfix/crowdsec
```

and try start/stop.

Thanks

**Comment by mmetc** [comment]
> Crowdsec made a barely visible effort here https://www.reddit.com/r/opnsense/comments/1fpt7xn/comment/lp3lixx/ -- but it would be nice to have someone on GitHub like in the old days ;)

Sorry for the lack of communication, I was not responsive enough across the board but you are right, github is the place I prefer as well. The issue page in crowdsecurity/crowdsec ensures attention from all the team.

By the way I'm thinking that a beta-test process is in order, because the plugin needs an overhaul and its user base has doubled in a few months.

**Comment by RudiKlein** [comment]
@mmetc I only observed the Crowdsec service (processes) hanging during system updates.

I have installed the hotfix. 
Tried to stop the crowdsec processes (to make sure they were using the hotfix), but one of them didn't wanna leave.
Rebooted the system 
Stopped and started crowdsec from the gui. That seemed to work.
Tried to check the process status from the CLI, but now SSH is giving me a hard time. (it's one of these days).

So, status unknown for now. I'll give it a go tomorrow. Today the users can't handle more downtime.

Rudi

**Comment by daygle** [comment]
I would like to share my testing related to the original issue.

I have found the below during testing: -

1. After a restart of OPNsense (but stopping Crowdsec before the restart). Then wait for OPNsense to boot, if you then perform a reboot it will wait for Crowdsec PIDS to stop (without restarting).
2. From the shell if you reboot, then receive the 'Waiting for PIDs' message select Ctrl+C then reboot again - it will reboot normally the second time.
3. Now the interesting thing is. If you, let's say restart OPNsense. Then, after OPNsense has started restart Crowdsec from the 'Services' section then reboot again the reboot will perform normally.
4. This can be replicated without the restart of OPNsense by running a shell and running 'service crowdsec stop' - you will receive the waiting for PIDs message. Please note point 1 - only after a reboot/without the Crowdsec services stopped/restarted.

**Comment by nhatlinh1982** [comment]
As i said before I also was having problem with this issue but I have another issue. While I am not sure if it is related to crowdsec but I was start having random shutdown of OPNsense VM in Proxmox AND having reboot problem. Because I only recently installed it crowdsec plugin, AND updated to OPNsense 24.7 from 24.1. I rolled back to old backup and currently running 24.7 OPNsense without crowdsec with NO issue for 2 days. I will install crowdsec after a week if there are no random reboot, and observe if crowdsec is causing this issue again.

**Comment by daygle** [comment]
@nhatlinh1982 Sounds like the same issue to me.

**Comment by mmetc** [comment]
> I would like to share my testing related to the original issue.
> 
> I have found the below during testing: -
[...]

Thanks for testing! You can update the script with this command until 1.6.3-2 is out.

```
# fetch -o /usr/local/etc/rc.d/crowdsec https://github.com/crowdsecurity/plugins/releases/download/crowdsec-1.6.3-2-hotfix/crowdsec
```

**Comment by daygle** [comment]
> > I would like to share my testing related to the original issue.
> > I have found the below during testing: -
> > [...]
> 
> Thanks for testing! You can update the script with this command until 1.6.3-2 is out.
> 
> ```
> # fetch -o /usr/local/etc/rc.d/crowdsec https://github.com/crowdsecurity/plugins/releases/download/crowdsec-1.6.3-2-hotfix/crowdsec
> ```

Thanks @mmetc - Great news, initial testing is showing that this fix has corrected the issue. Any chance you can have a look at the other issue with Crowdsec? Stop/Start button issue - https://github.com/opnsense/plugins/issues/4280.

The system will reboot. Do you want to proceed? [y/N]: y

>>> Invoking stop script 'beep'
>>> Invoking stop script 'freebsd'
Stopping crowdsec_firewall.
Stopping php_fpm.
Waiting for PIDS: 20611.
Stopping crowdsec.
Waiting for PIDS: 77982.
Waiting for PIDS: 78051.
Stopping nginx.
Waiting for PIDS: 13630.
Stopping acme_http_challenge.
Waiting for PIDS: 3552.
Stopping suricata.
Waiting for PIDS: 99017.
Stopping flowd.
Waiting for PIDS: 93666 94298.
Stopping flowd_aggregate...done
Stopping monit.
Waiting for PIDS: 85721.
crowdsec not running? (check /var/run/crowdsec_daemon.pid).
crowdsec_firewall is not running.
INFO/keactrl: Stopping kea-dhcp4...
INFO/keactrl: kea-dhcp6 isn't running.
INFO/keactrl: kea-dhcp-ddns isn't running.
INFO/keactrl: kea-ctrl-agent isn't running.
>>> Invoking stop script 'backup'
>>> Invoking backup script 'captiveportal'
>>> Invoking backup script 'dhcpleases'
>>> Invoking backup script 'duid'
>>> Invoking backup script 'netflow'
>>> Invoking backup script 'rrd'
>>> Invoking stop script 'config'
shutdown: [pid 91515]
Shutdown NOW!

*** FINAL System shutdown message from root@xxx ***

System going down IMMEDIATELY

**Comment by mmetc** [comment]
> > > I would like to share my testing related to the original issue.
> > > I have found the below during testing: -
> > > [...]
> > 
> > 
> > Thanks for testing! You can update the script with this command until 1.6.3-2 is out.
> > ```
> > # fetch -o /usr/local/etc/rc.d/crowdsec https://github.com/crowdsecurity/plugins/releases/download/crowdsec-1.6.3-2-hotfix/crowdsec
> > ```
> 
> Thanks @mmetc - Great news, initial testing is showing that this fix has corrected the issue. Any chance you can have a look at the other issue with Crowdsec? Stop/Start button issue - #4280.

Yes, I expect the start/stop button to work after the fetch. If it doesn't, I can inspect the crowdsec logs for any other issue. Run "cscli support dump" and send the file to support@crowdsec.net (partial configuration is included, without password/api keys of course)

**Comment by daygle** [comment]
Thanks again @mmetc.

Just letting you know that the 'Play' button remains green after CrowdSec is stopped - still an issue I am afraid. I'll send the logs to that email address.

**Comment by mmetc** [comment]
Seen and replied. As I wrote in the other issue, run "killall crowdsec" after the fetch to make sure there's no orphan process.

**Comment by RudiKlein** [comment]
@mmetc I have reproduced the issue I still have after installing the hotfix.

Hotfix installed
```$ killall crowdsec``` didn't kill the crowdsec processes.
```$ reboot``` to the rescue.

__*Status just after boot*__

    0 28250     1 0  68  0   12732    2272 kqread   Is    -    0:00.00 daemon: crowdsec[28405] (daemon)
    0 28405 28250 0  63  0 1482580  162092 kqread   I     -    0:32.90 /usr/local/bin/crowdsec -c /usr/local/etc/crowdsec/config.yaml
    0 37953 37094 1  20  0 1239152   44728 uwait    S     -    0:01.35 /usr/local/bin/crowdsec-firewall-bouncer -c /usr/local/etc/crowdsec/bouncers/crowdsec-firewall-bouncer.yaml -v (crowdsec-firewall-b)

__*Status after STOP from Services page*__ (won't go to STOPPED state.)

    0  1208   263 1  20  0   92144   64120 select   S     -    0:00.73 /usr/local/bin/php /usr/local/sbin/pluginctl -s crowdsec stop
    0  1986   263 1  68  0   13280    3056 wait     I     -    0:00.01 /bin/sh /usr/local/etc/rc.d/oscrowdsec stop
    0  2520  1986 1  68  0   13280    3172 wait     I     -    0:00.01 /bin/sh /usr/local/etc/rc.d/crowdsec stop

__*Status after ```$ killall crowdsec```*__

    0  1986   263 1  68  0   13280    3056 wait     I     -    0:00.01 /bin/sh /usr/local/etc/rc.d/oscrowdsec stop
    0  2520  1986 1  68  0   13280    3172 wait     I     -    0:00.01 /bin/sh /usr/local/etc/rc.d/crowdsec stop
    0 28405     1 2  20  0 1482580  163008 uwait    S     -    0:36.23 /usr/local/bin/crowdsec -c /usr/local/etc/crowdsec/config.yaml

I then used ```$ killall -9 crowdsec``` to kill the crowdsec processes succesfully.
After that stopping/starting from the Services page worked fine.

It seems there's still an issue after the hotfix.

Rudi

**Comment by daygle** [comment]
@RudiKlein It does awfully sound the same as what I was experiencing before the hotfix.

These are the steps I followed:-

1. Stopped Crowdsec via the services 'Stop' button.
2. Applied fix.
3. Rebooted OPNsense.
4. Stopped Crowdsec via the services 'Stop' button.
5. Killed Crowdsec via shell command.
6. Changed Crowdsec port from 8080 to 8085.
7. Started Crowdsec via the services 'Play' button.

I then was able to stop/restart Crowdsec via services and perform reboots without any issues so far.

**Comment by mmetc** [comment]
> @mmetc I have reproduced the issue I still have after installing the hotfix.

> I then used `$ killall -9 crowdsec` to kill the crowdsec processes succesfully. After that stopping/starting from the Services page worked fine.
> 
> It seems there's still an issue after the hotfix.

I changed the recommendation to "killall -9", that should do it, thanks!

**Comment by LaurenceJJones** [comment]
To keep thread updated `24.7.6` has just been released which includes the proper package @mmetc described

refs:
https://discourse.crowdsec.net/t/bug-opnsense-24-7-5-crowdsec-1-6-3/2057
https://forum.opnsense.org/index.php?topic=43305.0

**Comment by fichtner** [comment]
@mmetc @LaurenceJJones thanks guys, closing this then :)

**Comment by fichtner** [comment]
(if that was too soon we can reopen. getting mixed signals on reddit.)

**Comment by LaurenceJJones** [comment]
> (if that was too soon we can reopen. getting mixed signals on reddit.)

I believe the issue is some processes are still not getting killed cause the patch is only taking effect once the currently running PID is killed hence why in our thread we mention there **has to be manual intervention**. Once the original PID is killed the new patch takes effect and should have no issues.

**Comment by fichtner** [comment]
Judging purely by https://github.com/opnsense/ports/commit/4bd513d58 which gets installed prior to reboot it should fix the impending reboot as the updated script is called to stop it. That is, of course, under the assumption the fix is accurate.


Cheers,
Franco

**Comment by fichtner** [comment]
Ok so the issue is users having attempted the 24.7.5 update having the process stuck and will then continue to be stuck if attempting to go to 24.7.6 directly from the bad 24.7.5 state. But any forced reboot or the mentioned workaround would make it unstuck. Understood now.

**Comment by GitTimeraider** [comment]
Hmm.. had it last updates and have it again now. Kinda annoying having to manually intervene each time.

**Comment by LaurenceJJones** [comment]
> Hmm.. had it last updates and have it again now. Kinda annoying having to manually intervene each time.

Yes this patch fixes it so you will not have to manually intervene moving forward.

**Comment by LowHanger** [comment]
For what it's worth, I just had the same issue when upgrading from 24.5_3 to 24.6.  I ssh'ed and killed the PID which was stuck.  When I did that the install process on the GUI continued on as normal.

**Comment by julsssark** [comment]
Thank you for fixing this!

**Comment by roens** [comment]
Nice work folks!

I just ran an upgrade from `24.7.5_3` to `24.7.6`.

Of course it did hang on waiting for CrowdSec to quit, and I had to `killall -9 crowdsec`. It did then proceed with the remainder of the upgrade, then rebooted.

Looking forward to future unfettered reboots.

**Comment by mmetc** [comment]
@daygle can you please run "cscli support dump" and send the output to support@crowdsec.net? Thanks

**Comment by daygle** [comment]
> @daygle can you please run "cscli support dump" and send the output to [support@crowdsec.net](mailto:support@crowdsec.net)? Thanks

Thanks @mmetc - sent you an email and support logs.

**Comment by benisdev-py** [comment]
same problem still every update...

**Comment by julsssark** [comment]
I upgraded to 24.7.8. from 24.7.7 and the update worked without a problem.

**Comment by benisdev-py** [comment]
Oh okay did you have installed also crowdsec firewall bouncer?

**Comment by julsssark** [comment]
Yes. Running Crowdsec was installed and bouncer was running. What version of opnsense were you upgrading from/to?

**Comment by benisdev-py** [comment]
I also updater from 24.7.6 to a version => 24.7.6

**Comment by Starkstromkonsument** [comment]
The error still occurs on my instance (updated from 24.7.10_2 --> 24.7.11_2). My Instance is configured to reboot after every update. It has been rebooted several times since the patch.

It stops here:
```
>>> Invoking stop script 'beep'
>>> Invoking stop script 'freebsd'
Stopping crowdsec_firewall.
Stopping crowdsec.
Waiting for PIDS: 71801.
Waiting for PIDS: 4718
```

If I open a second terminal via ssh and do a `killall -9 crowdsec` there, the shutdown proceeds without further issues. Any hints where to search for the error?

**Comment by benisdev-py** [comment]
> The error still occurs on my instance (updated from 24.7.10_2 --> 24.7.11_2). My Instance is configured to reboot after every update. It has been rebooted several times since the patch.
> 
> It stops here:
> 
> ```
> >>> Invoking stop script 'beep'
> >>> Invoking stop script 'freebsd'
> Stopping crowdsec_firewall.
> Stopping crowdsec.
> Waiting for PIDS: 71801.
> Waiting for PIDS: 4718
> ```
> 
> If I open a second terminal via ssh and do a `killall -9 crowdsec` there, the shutdown proceeds without further issues. Any hints where to search for the error?

same…

**Comment by prime0n** [comment]
Same behaviour on version 25.1.1

**Comment by julsssark** [comment]
My upgrade from 25.1.0 to 25.1.1 worked without incident.

**Comment by RudiKlein** [comment]
I just went from 24.latest to 25.1.1, and the wait hang was there. Kill -9 moved it forward.
I like the QoL improvement showing the PID's.

SSH is a must have if this issue isn't resolved.
